keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,722 @@
+Node IP". Kubelet itself does not consider this a fatal error, but
+other code (like some e2e tests) does.)
+
+On clusters using an external cloud provider, `Node.Status.Addresses`
+is set by the cloud provider with no input from kubelet. (`--node-ip`
+has essentially no effect when using an external cloud provider.) In
+practice, this means that external cloud providers must always return
+IPv4 addresses first, since otherwise single-stack IPv4 clusters would
+not work correctly. In turn, that means that single-stack IPv6
+clusters _won't_ work correctly with external cloud providers.
+
+When using a legacy cloud provider, the same list of
+cloud-provider-provided addresses is used as a starting point, but:
+
+  1. If the user passes `--node-ip IP` to kubelet, and the provided
+     IP is one of the addresses returned by the cloud provider, then
+     all other IPs of the same `v1.NodeAddressType` are removed from
+     `Node.Status.Addresses`. (If the provided IP is not found, then
+     kubelet errors out.)
+
+  2. Otherwise, kubelet will sort the cloud-provider-provided
+     addresses by IP family (with IPv6 addresses first if the user
+     passed `--node-ip ::`, and with IPv4 addresses first
+     otherwise). Assuming that the cloud returned at least one IP of
+     the intended family, this will result in that becoming the
+     primary Node IP. However, if there are no IPs of the intended
+     family, it will silently fall back to using an IP of the other
+     family if one is available.
+
+On bare metal, if an explicit `--node-ip` has been passed, then
+kubelet will create a single `InternalIP` address with that IP.
+Otherwise several other options are tried (such as DNS, or picking an
+IP from the default network interface). If `--node-ip ::` was passed,
+then IPv6 DNS/interface addresses will be preferred; otherwise IPv4
+addresses will be preferred.
+
+#### New: `Node.Status.NodeIPs` and `--node-ips`
+
+To allow more consistent behavior without breaking any existing users,
+we will deprecate `--node-ip`, and deprecate the use of
+`Node.Status.Addresses` for determining the "primary" node IP. A new
+`--node-ips` argument will replace `--node-ip`, and a new
+`Node.Status.NodeIPs` field will provide an explicit indication of
+primary node IPs for components that were previously looking at
+`Node.Status.Addresses`.
+
+The new `--node-ips` argument can be either a single element
+describing the node IP configuration for a single family (indicating a
+mandatorily-single-stack node), or else a pair of elements describing
+both the IPv4 and IPv6 configurations. An element can be:
+
+  - a specific IP address, eg, `10.0.0.1` or `fd01::1`, indicating
+    that the node _must_ have that IP address for that IP family
+
+  - the string "`ipv4?`" or "`ipv6?`", indicating that the node _may_
+    have an IP address of that family, but is not required to
+
+  - the string "`ipv4`" or "`ipv6`", indicating that the node _must_
```

Comment:
I can't see the benefit of enforcing the ipFamily but I recognize that I need more time to digest the whole logic ... is not simpler going with only _may_? my reading is that this _must_ is mainly for making kubelet fail

Reply:
I really only put the "may" versions there to support the default behavior of "I didn't bother to tell you what kind of cluster I want, just read my mind and do the right thing". I don't think there's really a good use case for "I want this node to be single-stack IPv4, but if something goes wrong, feel free to just fall back to single-stack IPv6 without telling me". Except of course that we need to support that for backward compatibility with `--node-ip`.

Maybe I should make it impossible to explicitly specify the "may" versions on the command line, but still preserve the functionality internally for the default value and the `--node-ip` compatibility values.